builtin/common/vector.lua

```diff
@@ -122,3 +122,77 @@ function vector.sort(a, b)
 	return {x = math.min(a.x, b.x), y = math.min(a.y, b.y), z = math.min(a.z, b.z)},
 		{x = math.max(a.x, b.x), y = math.max(a.y, b.y), z = math.max(a.z, b.z)}
 end
+
+-- returns a new normalized 2d vector with random values
+function vector.random_2d()
+	local phi = math.random() * 2 * math.pi
+	return {
+		x = math.cos(phi),
+		y = math.sin(phi),
+		z = 0,
+	}
+end
+
+-- returns a new normalized 3d vector with random values
+function vector.random_3d()
+	local phi = math.random() * 2 * math.pi
+	local y = math.random() * 2 - 1
+	local r = math.sqrt(1 - y * y)
+	local x = r * math.cos(phi)
+	local z = r * math.sin(phi)
+	return {x = x, y = y, z = z}
+end
+
+-- returns a vector with a specific length
+function vector.set_length(v, l)
+	v = vector.normalize(v)
+	return vector.multiply(v, l)
+end
+
+-- limits the length of a vector
+function vector.limit(v, l)
+	local length = vector.length(v)
+	if length > l then
+		return vector.multiply(v, l / length)
+	end
+	return v
+end
+
+-- returns the dot product of two vectors
+function vector.dot(a, b)
+	return a.x * b.x + a.y * b.y + a.z * b.z
+end
+
+-- returns the cross product of two vectors
+function vector.cross(a, b)
+	return {x = a.y * b.z - a.z * b.y,
+		y = a.z * b.x - a.x * b.z,
+		z = a.x * b.y - a.y * b.x}
+end
+
+-- returns the scalar triple product of three vectors
+function vector.scalar_triple(a, b, c)
+	return vector.dot(a, vector.cross(b, c))
+end
+
+-- returns the vector triple product of three vectors
+function vector.vector_triple(a, b, c)
+	return vector.cross(a, vector.cross(b, c))
+end
+
+-- returns the angle between two 2d vectors (x,z) in radians
+-- y from a 3d vector is ignored
+function vector.angle_between_2d(a, b)
+	return math.atan2(a.x * b.z - a.z * b.x, a.x * b.x + a.z * b.z)
+end
```

Comment:
What about
```
function vector.angle_between(a, b)
	return math.acos(vector.dot(a, b) / vector.length(a) / vector.length(b))
end
```
?
(https://en.wikipedia.org/wiki/Dot_product#/media/File:Inner-product-angle.svg)

Reply:
I’d prefer this to be called `angle_between`; the original function does a different thing: it returns an *oriented* angle between projections on the XZ plane.

Reply:
Your link goes to your request of renaming the old angle_between to …_2d.
I've seen that HybridDog has suggested the same already in the old PR. https://github.com/minetest/minetest/pull/4277#discussion_r75578843
But all those comments are marked as outdated, I think, my new request here doesn't hurt.